### PR TITLE
Fuzzing updates: all-clusters-app

### PR DESCRIPTION
**Pull request by zcduthie**
As mentioned in [another pull request](https://github.com/project-chip/connectedhomeip/pull/27857), a colleague and I have completed some research investigating fuzzing.
This pull request summarises some of our findings that enable deeper fuzzing of the all-clusters-app fuzz target.
We note that the all-clusters-app target is not currently added to OSS-Fuzz integration - and most of the proposed changes below are more severe code changes (which may limit their appeal). Nonetheless, we hope at the very least this record is useful for future fuzzing efforts.

The below changes resulted in an additional ~5000 lines (45%) of all-clusters fuzzing coverage; notably including the addition of successful parsing of app and interaction model commands (+350% coverage of src/app).

### all-clusters-app-fuzzing
- Add dictionary (re-use TLV dictionary from unit tests)
- Add seed corpus (manually captured commissioning + door lock/unlock sequence using chip-tool and all-clusters-app)

Other:
- Add `CHIP_CONFIG_SECURITY_FUZZ_MODE` to *disable* encryption and signature verification
- Add fuzz driver manual injection of test (`sessionId=1`) session (and fallback to use this for secure unicast messages)
- Disable duplicate message check
- Perform fuzzing on an already commissioned all-clusters-app. For example by copying a pre-commissioned (with FUZZ_MODE above turned on) set of `/tmp/chip_kvs` credentials over prior to commencing fuzzing.

For example:

```
# OSS-Fuzz projects/connectedhomeip/build.sh
# ...
# ...

cp integrations/fuzz/precommissioned_credentials/* /tmp
```

### CODE REVIEW

File: config/standalone/CHIPProjectConfig.h [hunk 1]
```diff
@@ -56,6 +56,16 @@
...
 
+// For convenience, Chip Security Fuzz Mode can be enabled to
+// create deterministic behaviour to better facilitate fuzzing.
+//
+//    WARNING: This option circumvents basic Chip security functionality,
+//    including message encryption. Because of this it MUST NEVER BE ENABLED IN PRODUCTION BUILDS.
+//
+//    To build with this flag, pass 'treat_warnings_as_errors=false' to gn/ninja.
+//
+#define CHIP_CONFIG_SECURITY_FUZZ_MODE 0
```
Comment (bzbarsky-apple) [inline config/standalone/CHIPProjectConfig.h:67]: Shouldn't this be the default value, though?  And if so, why are we adding it to all the project configs?

File: src/transport/SecureMessageCodec.cpp [hunk 2]
```diff
@@ -48,6 +48,16 @@ CHIP_ERROR Encrypt(const CryptoContext & context, CryptoContext::ConstNonceView
 
     ReturnErrorOnFailure(payloadHeader.EncodeBeforeData(msgBuf));
 
+// Skip encryption and message integrity!
+#if CHIP_CONFIG_SECURITY_FUZZ_MODE
+#warning                                                                                                                           \
+    "Warning: CHIP_CONFIG_SECURITY_FUZZ_MODE=1 bypassing encryption! Node can only communicate with other nodes built with this flag set. Requires build flag 'treat_warnings_as_errors=false'."
```
Comment (bzbarsky-apple) [inline src/transport/SecureMessageCodec.cpp:54]: The "requires build flag" bit is:
1. Not really in need of being documented here.
2. Seems to not be needed at all, if this code just has `#else` around the bits we want to turn off....

Same for Decrypt.

File: src/transport/SessionManager.cpp [hunk 3]
```diff
@@ -739,8 +750,15 @@ void SessionManager::SecureUnicastMessageDispatch(const PacketHeader & partialPa
                       "Received a duplicate message with MessageCounter:" ChipLogFormatMessageCounter
                       " on exchange " ChipLogFormatExchangeId,
                       packetHeader.GetMessageCounter(), ChipLogValueExchangeIdFromReceivedHeader(payloadHeader));
+
+#if CHIP_CONFIG_SECURITY_FUZZ_MODE
+#warning "Warning: CHIP_CONFIG_SECURITY_FUZZ_MODE=1 bypassing duplicate message check!"
```
Comment (bzbarsky-apple) [inline src/transport/SessionManager.cpp:755]: This could use some documentation explaining why we would want to do that.

**Comment by zcduthie** [comment]
Our usage of `CHIP_CONFIG_SECURITY_FUZZ_MODE` was modelled on the existing `CHIP_CONFIG_SECURITY_TEST_MODE` flag. If this fuzzing approach is deemed a worthwhile inclusion - there are definitely improvements that can be made to the integration of this flag - especially proper integration into build tools to incorporate it as a flag such as is done with `oss_fuzz=true` in https://github.com/project-chip/connectedhomeip/pull/26010.
For our research we manually forced it on by modifying files that be seen in https://github.com/project-chip/connectedhomeip/pull/27858/commits/a01dfd1f942aefb6635d60c418dc7cffb0b2814b.

**Comment by andy31415** [comment]
PR not updated in >1 year and has merge conflicts. Closing as stale.

**Pull request closed without merging**